src/Pact/Types/SizeOf.hs

```diff
@@ -134,3 +147,95 @@ instance SizeOf UTCTime where
   -- Internally 'UTCTime' is just a 64-bit count of 'microseconds'
   sizeOf ti =
     (constructorCost 1) + (sizeOf (toPosixTimestampMicros ti))
+
+instance SizeOf Bool where
+  sizeOf _ = 0
+
+instance SizeOf () where
+  sizeOf _ = 0
+
+-- See: http://wiki.haskell.org/GHC/Memory_Footprint near bottom
+-- as well as https://blog.johantibell.com/2011/06/memory-footprints-of-some-common-data.html
+-- for both hash sets and hashmaps.
+instance (SizeOf k, SizeOf v) => SizeOf (HM.HashMap k v) where
+  sizeOf m = hmSize
+    where
+    hmSize = (5*hmLength + 4*(hmLength-1))*wordSize + contentLength
+    contentLength = HM.foldlWithKey' (\a k v -> a + sizeOf k + sizeOf v) 0 m
+    hmLength = fromIntegral (HM.size m)
+
+-- Note: Atm hashset is only a newtype wrapper over hashmap with unit as the element
+-- member of every entry. This means you don't pay at all for the cost of (),
+-- but you do pay for the extra constructor field of holding it, hence the `hsSize` bit
+-- stays roughly the same.
+instance (SizeOf k) => SizeOf (HS.HashSet k) where
+  sizeOf hs = hsSize
+    where
+    hsSize = (5*hsLength + 4*(hsLength-1))*wordSize + contentLength
+    contentLength = HS.foldl' (\a v -> a + sizeOf v) 0 hs
+    hsLength = fromIntegral (HS.size hs)
+
+instance (SizeOf a, SizeOf b) => SizeOf (Either a b)
+
+instance  (SizeOf a) => SizeOf (NE.NonEmpty a) where
+  sizeOf (a NE.:| rest) =
+    constructorCost 2 + sizeOf a + sizeOf rest
+
+
+class SizeOf1 f where
+  sizeOf1 :: SizeOf a => f a -> Bytes
+
+instance (SizeOf a, SizeOf1 f, SizeOf b) => SizeOf (Var a (f b)) where
+  sizeOf = \case
+    B a -> sizeOf a
+    F a -> sizeOf1 a
+
+instance (SizeOf b, SizeOf a, SizeOf1 f) => SizeOf (Scope b f a) where
+  sizeOf = sizeOf1 . unscope
+
+-- Generic deriving
+class GSizeOf f where
```

Comment:
Are these used in the empty decls in `Term`? How do we know they're working?

Reply:
I can add some tests.